app/scripts/components/exploration/components/datasets/block-utils.ts

```diff
@@ -0,0 +1,81 @@
+import {
+  endOfDay,
+  endOfMonth,
+  endOfYear,
+  startOfDay,
+  startOfMonth,
+  startOfYear,
+} from 'date-fns';
+import { TimeDensity } from '$components/exploration/types.d.ts';
+
+/**
+ * Calculate the start and end of a block of time, given a date and a time
+ * density
+ * @param date Starting date
+ * @param timeDensity Dataset time density
+ *
+ * @returns Array of two dates, the first being the start of the block and the
+ * second being the end of the block.
+ */
+export function getBlockBoundaries(date: Date, timeDensity: TimeDensity) {
+  switch (timeDensity) {
+    case TimeDensity.MONTH:
+      return [startOfMonth(date), endOfMonth(date)];
+    case TimeDensity.YEAR:
+      return [startOfYear(date), endOfYear(date)];
+  }
+
+  return [startOfDay(date), endOfDay(date)];
+}
+
+/**
+ * Lumps blocks of time together if they are too small to be seen on the chart.
+ *
+ * @param params.domain The dataset domain from which to calculate the blocks
+ * @param params.xScaled The xScaled function from the chart
+ * @param params.timeDensity The dataset time density
+ * @param params.minBlockSize The minimum size of a block
+ *
+ * @returns The blocks and whether or not any lumping happened.
+ */
+export function lumpBlocks({ domain, xScaled, timeDensity, minBlockSize = 4 }) {
```

Comment:
Have we considered the "disjointed repetitive block" use case? (ie, every day of 2015 then every day of 2017). 
I think that's not allowed by the STAC API, but just checking.

Reply:
Not sure I understood what you mean here